rust/protocol/src/curve.rs

```diff
@@ -280,41 +347,118 @@ impl TryFrom<&[u8]> for PrivateKey {
     }
 }
 
-#[derive(Copy, Clone)]
+impl subtle::ConstantTimeEq for PrivateKey {
+    fn ct_eq(&self, other: &PrivateKey) -> subtle::Choice {
+        if self.key_type() != other.key_type() {
+            return 0.ct_eq(&1);
+        }
+        self.key_data().ct_eq(other.key_data())
+    }
+}
+
+impl PartialEq for PrivateKey {
+    fn eq(&self, other: &PrivateKey) -> bool {
+        bool::from(self.ct_eq(other))
+    }
+}
+
+impl Ord for PrivateKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.key_type() != other.key_type() {
+            return self.key_type().cmp(&other.key_type());
+        }
+        constant_time_cmp(self.key_data(), other.key_data())
+    }
+}
+
+impl PartialOrd for PrivateKey {
+    fn partial_cmp(&self, other: &PrivateKey) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl fmt::Debug for PrivateKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            // Do not print out private keys in debug logs.
+            "PrivateKey {{ key_type={:?}, serialize=<...> }}",
+            self.key_type(),
+        )
+    }
+}
+
+/// A matching public and private key pair which can be used to encrypt and sign messages.
+///
+///```
+/// # fn main() -> libsignal_protocol::error::Result<()> {
+/// use libsignal_protocol::{KeyPair, PrivateKey, PublicKey};
+/// use rand::rngs::OsRng;
+/// use std::convert::TryInto;
+///
+/// let alice = KeyPair::generate(&mut OsRng);
+/// assert!(alice == alice.clone());
+/// let bob = KeyPair::generate(&mut OsRng);
+/// assert!(alice != bob);
+///
+/// // The first byte of `alice_zero` is 0, less than `bob_one` with 1, lexicographically:
```

Comment:
I don't think this is a good example. For one thing, the first and last bytes of Curve25519 keys have restrictions (the ones enforced in deserialization). But more importantly, there doesn't need to be a guarantee of *how* the keys are ordered, only that they can be.

Reply:
…also, none of this has to do with KeyPair. Maybe it's module-level docs?

Reply:
I have removed this example, removed the `Ord` impl, and moved a smaller version of this doctest to the module-level docs!